File: src/main/java/org/theta4j/osc/OSCClient.java
/*
 * Copyright (C) 2019 theta4j project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.theta4j.osc;

import com.burgstaller.okhttp.AuthenticationCacheInterceptor;
import com.burgstaller.okhttp.CachingAuthenticatorDecorator;
import com.burgstaller.okhttp.digest.CachingAuthenticator;
import com.burgstaller.okhttp.digest.Credentials;
import com.burgstaller.okhttp.digest.DigestAuthenticator;
import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import okhttp3.*;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.io.IOException;
import java.io.Reader;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

import static java.net.HttpURLConnection.HTTP_UNAUTHORIZED;

/**
 * OSCClient is a client of Open Spherical Camera API.
 *
 * @see <a href="https://developers.google.com/streetview/open-spherical-camera/">Open Spherical Camera API</a>
 */
public final class OSCClient {
    private static final Gson GSON = new Gson();

    private static final Command<JsonObject, OptionSet> GET_OPTIONS_COMMAND = Command.create("camera.getOptions", JsonObject.class, OptionSet.class);

    private static final Command<OptionSet, Void> SET_OPTIONS_COMMAND = Command.create("camera.setOptions", OptionSet.class, Void.class);

    private final String endpoint;

    private final OkHttpClient httpClient;

    private OSCClient(final String endpoint, final OkHttpClient httpClient) {
        this.endpoint = endpoint;
        this.httpClient = httpClient;
    }

    /**
     * Create OSCClient.
     *
     * @param endpoint Endpoint of the OSC API server. For example "http://192.168.1.1".
     * @throws NullPointerException if endpoint is null.
     */
    public static OSCClient create(@Nonnull final String endpoint) {
        Objects.requireNonNull(endpoint, "endpoint can not be null.");

        return new OSCClient(endpoint, new OkHttpClient());
    }

    /**
     * Create OSCClient with Digest authentication.
     *
     * @param endpoint Endpoint of the OSC API server. For example "http://192.168.1.1".
     * @param username Username for Digest authentication.
     * @param password Password for Digest authentication.
     * @throws NullPointerException if endpoint, username or password is null.
     */
    public static OSCClient createWithDigestAuthentication(@Nonnull final String endpoint, @Nonnull final String username, @Nonnull final String password) {
        Objects.requireNonNull(endpoint, "endpoint can not be null.");
        Objects.requireNonNull(username, "username can not be null.");
        Objects.requireNonNull(password, "password can not be null.");

        final com.burgstaller.okhttp.digest.Credentials credentials = new Credentials(username, password);
        final DigestAuthenticator authenticator = new DigestAuthenticator(credentials);
        final Map<String, CachingAuthenticator> authCache = new ConcurrentHashMap<>();

        final OkHttpClient httpClient = new OkHttpClient.Builder()
                .authenticator(new CachingAuthenticatorDecorator(authenticator, authCache))
                .addInterceptor(new AuthenticationCacheInterceptor(authCache))
                .build();

        return new OSCClient(endpoint, httpClient);
    }

    /**
     * GET /osc/info
     *
     * @param resultType a type of the result value.
     * @return Information object
     * @throws IOException          I/O error is occurred.
     * @throws NullPointerException if resultType is null.
     * @see <a href="https://developers.google.com/streetview/open-spherical-camera/guides/osc/info">Info | Open Spherical Camera API</a>
     */
    @Nonnull
    public <T> T info(@Nonnull final Class<T> resultType) throws IOException {
        Objects.requireNonNull(resultType, "resultType can not be null.");

        final JsonObject response = httpGet(endpoint + "/osc/info");
        return GSON.fromJson(response, resultType);
    }

    /**
     * POST /osc/state
     *
     * @param resultType a type of the state value.
     * @throws IOException          I/O error is occurred.
     * @throws NullPointerException if resultType is null.
     * @see <a href="https://developers.google.com/streetview/open-spherical-camera/guides/osc/state">State | Open Spherical Camera API</a>
     */
    @Nonnull
    public <T> OSCState<T> state(@Nonnull final Class<T> resultType) throws IOException {
        Objects.requireNonNull(resultType, "resultType can not be null.");

        final JsonObject response = httpPost(endpoint + "/osc/state", null);
        return OSCState.valueOf(response, resultType);
    }

    /**
     * Check for the updates of the state.
     *
     * @param fingerprint fingerprint of the last state.
     * @return updated fingerprint.
     * @throws IOException          I/O error is occurred.
     * @throws OSCException         Server returned error response.
     * @throws NullPointerException if fingerprint is null.
     * @see OSCClient#state(Class)
     * @see <a href="https://developers.google.com/streetview/open-spherical-camera/guides/osc/checkforupdates">CheckForUpdates | Open Spherical Camera API</a>
     */
    @Nonnull
    public String checkForUpdates(@Nonnull final String fingerprint) throws IOException {
        Objects.requireNonNull(fingerprint, "fingerprint can not be null.");

        final JsonObject request = new JsonObject();
        request.addProperty("stateFingerprint", fingerprint);
        final JsonObject response = httpPost(endpoint + "/osc/checkForUpdates", request);

        if (Objects.equals(response.get("state").getAsString(), "error")) {
            throw GSON.fromJson(response, OSCException.class);
        }

        return response.get("stateFingerprint").getAsString();
    }

    /**
     * Execute command.
     *
     * @param command the command to execute.
     * @param <P>     a type of the parameter.
     * @param <R>     a type of the result.
     * @return The response of the command.
     * @throws IOException          I/O error is occurred.
     * @throws OSCException         Server returned error response.
     * @throws NullPointerException if command is null.
     */
    @Nonnull
    public <P, R> CommandResponse<R> commandExecute(@Nonnull final Command<P, R> command) throws IOException {
        return commandExecute(command, null);
    }

    /**
     * Execute command.
     *
     * @param command   the command to execute.
     * @param parameter the parameter of the command.
     * @param <P>       a type of the parameter.
     * @param <R>       a type of the result.
     * @return The response of the command.
     * @throws IOException          I/O error is occurred.
     * @throws OSCException         Server returned error response.
     * @throws NullPointerException if command is null.
     */
    @Nonnull
    public <P, R> CommandResponse<R> commandExecute(@Nonnull final Command<P, R> command, @Nullable final P parameter) throws IOException {
        Objects.requireNonNull(command, "command can not be null.");

        final CommandRequest reqBody = new CommandRequest(command.getName(), parameter);
        final JsonObject resBody = httpPost(endpoint + "/osc/commands/execute", GSON.toJsonTree(reqBody).getAsJsonObject());

        final CommandResponse<R> response = CommandResponse.valueOf(resBody, command.getResultType());

        if (response.getError() != null) {
            throw response.getError();
        }

        return response;
    }

    /**
     * Check status of the executed command.
     *
     * @param response The response value of last
     * @return Updated command response.
     * @throws IOException          I/O error is occurred.
     * @throws OSCException         Server returned error response.
     * @throws NullPointerException if response is null or response does not have ID.
     * @see <a href="https://developers.google.com/streetview/open-spherical-camera/guides/osc/commands/status">Status | Open Spherical Camera API</a>
     */
    @Nonnull
    public <R> CommandResponse<R> commandStatus(@Nonnull final CommandResponse<R> response) throws IOException {
        Objects.requireNonNull(response, "response can not be null.");

        // request and get response
        final JsonObject reqBody = new JsonObject();
        reqBody.addProperty("id", response.getID());
        final JsonObject resBody = httpPost(endpoint + "/osc/commands/status", reqBody);

        final CommandResponse<R> updatedResponse = CommandResponse.valueOf(resBody, response.getResultType());

        if (updatedResponse.getError() != null) {
            throw updatedResponse.getError();
        }

        return updatedResponse;
    }

    /**
     * Get option value.
     *
     * @param option Option to acquire value.
     * @param <T>    Type of option value.
     * @return Acquired option value.
     * @throws IOException          I/O error is occurred.
     * @throws OSCException         Server returned error response.
     * @throws NullPointerException if option is null.
     * @see <a href="https://developers.google.com/streetview/open-spherical-camera/reference/camera/getoptions">camera.getOptions | Open Spherical Camera API</a>
     */
    @Nonnull
    public <T> T getOption(@Nonnull final Option<T> option) throws IOException {
        Objects.requireNonNull(option, "options can not be null.");
        return getOptions(option).get(option);
    }

    /**
     * Get option value.
     *
     * @param option Option to acquire value.
     * @param <T>    Type of option value.
     * @return Acquired option values.
     * @throws IOException          I/O error is occurred.
     * @throws OSCException         Server returned error response.
     * @throws NullPointerException if option is null.
     * @see <a href="https://developers.google.com/streetview/open-spherical-camera/reference/camera/getoptions">camera.getOptions | Open Spherical Camera API</a>
     */
    @Nonnull
    public <T> List<T> getOption(@Nonnull final ArrayOption<T> option) throws IOException {
        Objects.requireNonNull(option, "options can not be null.");
        return getOptions(option).get(option);
    }

    /**
     * Get options.
     *
     * @param options options to acquire.
     * @return Acquired options values.
     * @throws IOException              I/O error is occurred.
     * @throws OSCException             Server returned error response.
     * @throws NullPointerException     if options is null or contains null.
     * @throws IllegalArgumentException if length of options is 0.
     * @see <a href="https://developers.google.com/streetview/open-spherical-camera/reference/camera/getoptions">camera.getOptions | Open Spherical Camera API</a>
     */
    @Nonnull
    public OptionSet getOptions(@Nonnull final Option... options) throws IOException {
        return getOptions(Arrays.asList(options));
    }

    /**
     * Get options.
     *
     * @param options options to acquire.
     * @return Acquired options values.
     * @throws IOException              I/O error is occurred.
     * @throws OSCException             Server returned error response.
     * @throws NullPointerException     if options is null or contains null.
     * @throws IllegalArgumentException if length of options is 0.
     * @see <a href="https://developers.google.com/streetview/open-spherical-camera/reference/camera/getoptions">camera.getOptions | Open Spherical Camera API</a>
     */
    @Nonnull
    public OptionSet getOptions(@Nonnull final Collection<Option> options) throws IOException {
        Objects.requireNonNull(options, "options can not be null.");
        if (options.size() == 0) {
            throw new IllegalArgumentException("options must have 1 or more entries.");
        }
        if (options.contains(null)) {
            throw new NullPointerException("names can not contain null.");
        }

        final JsonArray optionNames = new JsonArray(options.size());
        options.stream()
                .map(Option::getName)
                .forEach(optionNames::add);

        final JsonObject parameter = new JsonObject();
        parameter.add("optionNames", optionNames);

        final CommandResponse<OptionSet> response = commandExecute(GET_OPTIONS_COMMAND, parameter);

        if (response.getError() != null) {
            throw response.getError();
        }

        return response.getResult();
    }

    /**
     * Set options value.
     *
     * @param option Option to set value.
     * @param value  Option value to set.
     * @param <T>    Type of option value.
     * @throws IOException          I/O error is occurred.
     * @throws OSCException         Server returned error response.
     * @throws NullPointerException if option or value is null.
     */
    @Nonnull
    public <T> void setOption(@Nonnull final Option<T> option, T value) throws IOException {
        final OptionSet optionSet = new OptionSet.Builder()
                .put(option, value)
                .build();
        setOptions(optionSet);
    }

    /**
     * Set optionSet.
     *
     * @param optionSet OptionSet values to set.
     * @throws IOException          I/O error is occurred.
     * @throws OSCException         Server returned error response.
     * @throws NullPointerException if optionSet is null.
     * @see <a href="https://developers.google.com/streetview/open-spherical-camera/reference/camera/setoptions">camera.setOptions | Open Spherical Camera API</a>
     */
    public void setOptions(@Nonnull final OptionSet optionSet) throws IOException {
        Objects.requireNonNull(optionSet, "optionSet can not be null.");

        final CommandResponse<Void> response = commandExecute(SET_OPTIONS_COMMAND, optionSet);

        if (response.getError() != null) {
            throw response.getError();
        }
    }

    // Helpers

    private JsonObject httpGet(final String url) throws IOException {
        final Request request = new Request.Builder()
                .url(url)
                .get()
                .addHeader("Accept", "application/json")
                .addHeader("X-XSRF-Protected", "1")
                .build();

        final Response response = httpClient.newCall(request).execute();

        if (response.code() == HTTP_UNAUTHORIZED) {
            throw new IOException(response.message());
        }

        try (final Reader r = response.body().charStream()) {
            return GSON.fromJson(r, JsonObject.class);
        }
    }

    private JsonObject httpPost(final String url, final JsonObject body) throws IOException {
        final RequestBody requestBody = RequestBody.create(
                MediaType.parse("application/json; charset=UTF-8"),
                body != null ? GSON.toJson(body) : "");

        final Request request = new Request.Builder()
                .url(url)
                .post(requestBody)
                .addHeader("Accept", "application/json")
                .addHeader("X-XSRF-Protected", "1")
                .build();

        final Response response = httpClient.newCall(request).execute();

        if (response.code() == HTTP_UNAUTHORIZED) {
            throw new IOException(response.message());
        }

        try (final Reader r = response.body().charStream()) {
            return GSON.fromJson(r, JsonObject.class);
        }
    }
}

File: src/main/java/org/theta4j/osc/OSCClient.kt
/*
 * Copyright (C) 2022 theta4j project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.theta4j.osc

import com.burgstaller.okhttp.AuthenticationCacheInterceptor
import com.burgstaller.okhttp.CachingAuthenticatorDecorator
import com.burgstaller.okhttp.digest.CachingAuthenticator
import com.burgstaller.okhttp.digest.Credentials
import com.burgstaller.okhttp.digest.DigestAuthenticator
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import kotlinx.serialization.KSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.logging.HttpLoggingInterceptor
import java.io.IOException
import java.util.concurrent.ConcurrentHashMap

class OSCClient private constructor(
        private val endpoint: String,
        private val httpClient: OkHttpClient,
        private val streamHttpClient: OkHttpClient,
) {
    companion object {
        private val GET_OPTIONS_COMMAND = Command.create("camera.getOptions", GetOptions.Parameter.serializer(), GetOptions.Result.serializer())
        private val SET_OPTIONS_COMMAND = Command.create("camera.setOptions", SetOptions.Parameter.serializer(), Unit.serializer())

        private val format = Json { ignoreUnknownKeys = true }

        fun create(endpoint: String): OSCClient {
            val httpClient = OkHttpClient.Builder().apply {
                addInterceptor(HttpLoggingInterceptor(logger::debug).setLevel(HttpLoggingInterceptor.Level.BODY))
            }.build()
            val streamHttpClient = OkHttpClient.Builder().apply {
                addInterceptor(HttpLoggingInterceptor(logger::debug).setLevel(HttpLoggingInterceptor.Level.HEADERS))
            }.build()
            return OSCClient(endpoint, httpClient, streamHttpClient)
        }

        fun createWithDigestAuthentication(endpoint: String, username: String, password: String): OSCClient {
            val credentials = Credentials(username, password)
            val authenticator = DigestAuthenticator(credentials)
            val authCache = ConcurrentHashMap<String, CachingAuthenticator>()
            val httpClient = OkHttpClient.Builder().apply {
                authenticator(CachingAuthenticatorDecorator(authenticator, authCache))
                addInterceptor(AuthenticationCacheInterceptor(authCache))
                addInterceptor(HttpLoggingInterceptor(logger::debug).setLevel(HttpLoggingInterceptor.Level.BODY))
            }.build()
            val streamHttpClient = OkHttpClient.Builder().apply {
                authenticator(CachingAuthenticatorDecorator(authenticator, authCache))
                addInterceptor(AuthenticationCacheInterceptor(authCache))
                addInterceptor(HttpLoggingInterceptor(logger::debug).setLevel(HttpLoggingInterceptor.Level.HEADERS))
            }.build()
            return OSCClient(endpoint, httpClient, streamHttpClient)
        }
    }

    suspend fun <T> info(serializer: KSerializer<T>): T = format.decodeFromJsonElement(serializer, httpGet("/osc/info"))

    suspend fun <T> state(serializer: KSerializer<T>): OSCState<T> = OSCState.decode(serializer, httpPost("/osc/state"))

    suspend fun checkForUpdates(fingerprint: String): String {
        val reqBody = buildJsonObject { put("stateFingerprint", fingerprint) }
        val response = httpPost("/osc/checkForUpdates", reqBody)
        response["error"]?.let { throw format.decodeFromJsonElement<OSCException>(it) }
        return response["stateFingerprint"]!!.jsonPrimitive.content
    }

    suspend fun <P, R> commandExecute(command: Command<P, R>, parameters: P? = null): CommandResponse<R> {
        val reqBody = buildJsonObject {
            put("name", command.name)
            if (parameters != null) {
                put("parameters", Json.encodeToJsonElement(command.parameterSerializer, parameters))
            }
        }
        val resBody = httpPost("/osc/commands/execute", reqBody)
        val response = CommandResponse.decode(command.resultSerializer, resBody)
        response.error?.let { throw it }
        return response
    }

    suspend fun <R> commandStatus(response: CommandResponse<R>): CommandResponse<R> {
        val reqBody = buildJsonObject {
            put("id", response.id)
        }
        val resBody = httpPost("/osc/commands/status", reqBody)
        val newResponse = CommandResponse.decode(response.resultSerializer, resBody)
        newResponse.error?.let { throw it }
        return newResponse
    }

    suspend fun <R> await(response: CommandResponse<R>, interval: Long = 100): CommandResponse<R> {
        var res = response
        while (res.state == CommandState.IN_PROGRESS) {
            res = commandStatus(res)
            delay(interval)
        }
        return res
    }

    suspend fun <T> getOption(option: Option<T>): T = getOptions(option)[option]!!

    suspend fun <T> getOption(option: ArrayOption<T>): List<T> = getOptions(option)[option]!!

    suspend fun getOptions(vararg options: Option<*>): OptionSet = getOptions(listOf(*options))

    suspend fun getOptions(options: Collection<Option<*>>): OptionSet {
        require(options.isNotEmpty()) { "options must have 1 or more entries." }
        val optionNames = options.map { it.name }
        val parameter = GetOptions.Parameter(optionNames)
        val response = commandExecute(GET_OPTIONS_COMMAND, parameter)
        response.error?.let { throw it }
        return response.result!!.options
    }

    suspend fun <T> setOption(option: Option<T>, value: T) {
        setOptions {
            put(option, value)
        }
    }

    suspend fun setOptions(optionSet: OptionSet) {
        val parameter = SetOptions.Parameter(optionSet)
        val response = commandExecute(SET_OPTIONS_COMMAND, parameter)
        response.error?.let { throw it }
    }

    suspend fun setOptions(block: OptionSetBuilder.() -> Unit) {
        val optionSet = OptionSetBuilder().apply(block).build()
        setOptions(optionSet)
    }

    suspend fun getLivePreview(): MJpegInputStream {
        val reqBody = buildJsonObject {
            put("name", "camera.getLivePreview")
        }
        val contentType = "application/json; charset=UTF-8".toMediaType()
        val requestBody = Json.encodeToString(reqBody).toRequestBody(contentType)
        val request = Request.Builder()
                .url("$endpoint/osc/commands/execute")
                .post(requestBody)
                .addHeader("Accept", "application/json")
                .addHeader("X-XSRF-Protected", "1")
                .build()
        val response = withContext(Dispatchers.IO) { streamHttpClient.newCall(request).execute() }
        if (response.code == 401) throw IOException(response.message)
        if (response.code != 200) {
            val resBody = withContext(Dispatchers.IO) { response.body!!.string() }
            val jsonObject = format.decodeFromString(JsonObject.serializer(), resBody)
            val commandRes = CommandResponse.decode(Unit.serializer(), jsonObject)
            throw commandRes.error!!
        }
        return MJpegInputStream("---osclivepreview---", response.body!!.byteStream())
    }

    //
    // Helpers
    //

    private suspend fun httpGet(path: String): JsonObject {
        val request = Request.Builder()
                .url(endpoint + path)
                .get()
                .addHeader("Accept", "application/json")
                .addHeader("X-XSRF-Protected", "1")
                .build()
        val response = withContext(Dispatchers.IO) { httpClient.newCall(request).execute() }
        if (response.code == 401) throw IOException(response.message)
        val resBody = withContext(Dispatchers.IO) { response.body!!.string() }
        return Json.decodeFromString(JsonObject.serializer(), resBody)
    }

    private suspend fun httpPost(path: String, body: JsonObject? = null): JsonObject {
        val contentType = "application/json; charset=UTF-8".toMediaType()
        val requestBody = Json.encodeToString(body).toRequestBody(contentType)
        val request = Request.Builder()
                .url(endpoint + path)
                .post(requestBody)
                .addHeader("Accept", "application/json")
                .addHeader("X-XSRF-Protected", "1")
                .build()
        val response = withContext(Dispatchers.IO) { httpClient.newCall(request).execute() }
        if (response.code == 401) throw IOException(response.message)
        val resBody = withContext(Dispatchers.IO) { response.body!!.string() }
        return Json.decodeFromString(JsonObject.serializer(), resBody)
    }
}
